Pylontech integration (solar battery bank)

functionpointer opened the pull request:
# What does this implement/fix?

Reads state of solar battery banks made by Pylontech.

Supports state of charge (%), voltage (V), current (A) and more for several banks individually.

![](https://raw.githubusercontent.com/esphome/esphome-docs/a412821570d80191c9e2feb26078cb91ef56d305/components/images/pylontech.jpg)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** none

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2841

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
uart:
  - id: pylontech_uart
    tx_pin: GPIO1
    rx_pin: GPIO3
    baud_rate: 115200

pylontech:
  - uart_id: pylontech_uart
    id: pylontech0
    update_interval: 5s

sensor:
  - platform: pylontech
    pylontech_id: pylontech0
    battery: 1

    voltage:
      id: bat1_voltage
      name: "Battery1 Voltage"
    current:
      id: bat1_current
      name: "Battery1 Current"
    coulomb:
      id: bat1_soc
      name: "Battery1 State of Charge"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

functionpointer committed:
pylontech: new integration!

Reads state of battery banks made by Pylontech.

functionpointer committed:
pylontech: update codeowners

functionpointer committed:
pylontech: format with clang-format

functionpointer committed:
pylontech: more formatting

SeByDocKy commented:
Great adding !!!

functionpointer committed:
pylontech: don't clang-format the example

the parsing example consists of two long lines
clang-format wants to shorten them by introducing line breaks.
this makes the example hard to read and mostly useless.

functionpointer committed:
Merge branch 'dev' into pylontech

norcon commented:
Can't wait to use it :-) thanks a lot!

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
pylontech: change TAG to pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge remote-tracking branch 'upstream/dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Pylontech: fix test4

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

jesserockz commented on esphome/components/pylontech/__init__.py:
```diff
@@ -0,0 +1,77 @@
+    .extend(cv.polling_component_schema("60s"))
+    .extend(uart.UART_DEVICE_SCHEMA)
+)
+# Each instance of PylontechComponent may address multiple batteries
+# With their own sensors for voltage, current, etc
+# To make this possible, each PylontechComponent has an array for each type of sensor
+# Problem: we need to know how long those arrays should be
+# Solution: keep track of the maximum index that is used in the entire esphome config
+# This is done using the global _MAX_BATTERY_INDEX
+# Whenever a battery index is used, mark_battery_index_in_use is called to keep track
+# add_num_batteries() will be called after all configs are processed to tell the c code the final count
+_MAX_BATTERY_INDEX = -1
+
+
+async def mark_battery_index_in_use(index: int, parent):
+    global _MAX_BATTERY_INDEX
+    _MAX_BATTERY_INDEX = max(_MAX_BATTERY_INDEX, index)
+    cg.add(parent.mark_battery_index_in_use(index))
+
+
+@coroutine_with_priority(-5.0)
+async def add_num_batteries():
+    if not getattr(add_num_batteries, "already_ran", False):
+        if _MAX_BATTERY_INDEX < 0:
+            _LOGGER.error("no idea how many batteries!")
+        cg.add_define("PYLONTECH_NUM_BATTERIES", _MAX_BATTERY_INDEX)
+        add_num_batteries.already_ran = True
```
A much simpler method would just be to have `num_batteries` as a config item on the main component...

functionpointer replied:
Very much simpler, yes. I didn't like it because it would be redundant. [The config](https://deploy-preview-2841--esphome.netlify.app/components/pylontech#sensor) already requires a separate block for every battery.

All this code is really doing is counting how many config blocks are defined.
Is there a better way of doing this than what is currently implemented?

functionpointer replied:
Well, apparently not.

Do you want me to implement `num_batteries`? I don't like it, but i will implement it to get this PR merged.

functionpointer replied:
Implemented it.

functionpointer replied:
Turns out the recently merged ld2420 uses `std::vector`. I did not know this was allowed in embedded context.
So i rewrote this PR with a `std::vector` at its core, which solves all the problems i had.

No crazy validation, no global variables, no define. So much cleaner IMO

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge remote-tracking branch 'upstream/dev' into pylontech-wip

functionpointer committed:
pylontech: Use explicit num_batteries instead of auto-detecting it

this greatly simpifies the code

functionpointer committed:
remove accidentally added test

functionpointer committed:
pylontech: define the global variable

functionpointer committed:
Merge branch 'dev' into pylontech

jesserockz commented on esphome/components/pylontech/sensor/__init__.py:
```diff
@@ -0,0 +1,92 @@
+async def to_code(config):
+    paren = await cg.get_variable(config[CONF_PYLONTECH_ID])
+    bat: int = config[CONF_BATTERY]
+    await check_battery_index(bat)
```
Validation is not to be performed inside `to_code`

Please see other components implementations of `FINAL_VALIDATE_SCHEMA` to validate sensor configs against parent configs.
```suggestion
```

functionpointer replied:
Good to know `FINAL_VALIDATE_SCHEMA` exists!

jesserockz commented on esphome/components/pylontech/pylontech.h:
```diff
@@ -0,0 +1,89 @@
+
+class PylontechComponent : public PollingComponent, public uart::UARTDevice {
+ public:
+  PYLONTECH_SENSOR(voltage)
+  PYLONTECH_SENSOR(current)
+  PYLONTECH_SENSOR(temperature)
+  PYLONTECH_SENSOR(temperature_low)
+  PYLONTECH_SENSOR(temperature_high)
+  PYLONTECH_SENSOR(voltage_low)
+  PYLONTECH_SENSOR(voltage_high)
+
+  PYLONTECH_SENSOR(coulomb)
+  PYLONTECH_SENSOR(mos_temperature)
```
All sensor code needs to be wrapped similar to text sensors.
```suggestion
#ifdef USE_SENSOR
  PYLONTECH_SENSOR(voltage)
  PYLONTECH_SENSOR(current)
  PYLONTECH_SENSOR(temperature)
  PYLONTECH_SENSOR(temperature_low)
  PYLONTECH_SENSOR(temperature_high)
  PYLONTECH_SENSOR(voltage_low)
  PYLONTECH_SENSOR(voltage_high)

  PYLONTECH_SENSOR(coulomb)
  PYLONTECH_SENSOR(mos_temperature)
#endif
```

jesserockz commented on esphome/components/pylontech/__init__.py:
```diff
@@ -0,0 +1,68 @@
+
+    global num_batteries
+    num_batteries = config[CONF_NUM_BATTERIES]
+    cg.add_define("PYLONTECH_NUM_BATTERIES", config[CONF_NUM_BATTERIES])
```
To follow the format we use in ESPHome, the define should be like this:
```suggestion
    cg.add_define("USE_PYLONTECH_NUM_BATTERIES", config[CONF_NUM_BATTERIES])
```

functionpointer replied:
Removed entirely :)

jesserockz commented on esphome/components/pylontech/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    .extend(uart.UART_DEVICE_SCHEMA)
+)
+
+num_batteries = -1
+
+
+async def check_battery_index(index: int):
+    if not (1 <= index <= num_batteries):
+        raise cv.Invalid(f"battery expected between 1 and {num_batteries}")
```
See my other comment on validation against parent

```suggestion
```

jesserockz commented on esphome/components/pylontech/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    await cg.register_component(var, config)
+    await uart.register_uart_device(var, config)
+
+    global num_batteries
+    num_batteries = config[CONF_NUM_BATTERIES]
```
```suggestion
```

jesserockz commented on esphome/components/pylontech/pylontech.cpp:
```diff
@@ -0,0 +1,174 @@
+
+  for (int bat_num = 0; bat_num < this->max_battery_index_; bat_num++) {
+    ESP_LOGCONFIG(TAG, " Battery %d", bat_num + 1);
+    if (this->voltage_[bat_num]) {
+      LOG_SENSOR("  ", "Voltage", this->voltage_[bat_num]);
+    }
```
The LOG_SENSOR macro already does a nullptr check
```suggestion
    LOG_SENSOR("  ", "Voltage", this->voltage_[bat_num]);
```

jesserockz commented on esphome/components/pylontech/pylontech.cpp:
```diff
@@ -0,0 +1,174 @@
+  bat_num--;
+
+  if (this->voltage_[bat_num]) {
+    this->voltage_[bat_num]->publish_state(((float) volt) / 1000);
```
```suggestion
    this->voltage_[bat_num]->publish_state(((float) volt) / 1000.0f);
```

jesserockz commented on esphome/components/pylontech/pylontech.h:
```diff
@@ -0,0 +1,89 @@
+  int buffer_index_read_ = 0;
+};
+
+class PylontechTextComponent : public Component {
+ public:
+  PylontechTextComponent(PylontechComponent *parent) {}
+};
```
Not used?

functionpointer replied:
Indeed. Removed.

functionpointer committed:
pylontech: use FINAL_VALIDATE

functionpointer committed:
pylontech: refactor to eliminate NUM_BATTERIES

Looking at ld2420 it turns out std::vector can be used!
This solves the main reason why NUM_BATTERIES was introduced.

The new implementation uses no crazy stuff at all :)
Just one std::vector.

functionpointer committed:
pylontech: formatting

functionpointer committed:
pylontech: more formatting

functionpointer committed:
pylontech: fix null checks

functionpointer committed:
pylontech: fix current_state text sensor

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer commented:
@jesserockz i addressed all issues, this code is ready for another review.

I have since also tested it with a US3000C, and it works great.

norcon commented:
> @jesserockz i addressed all issues, this code is ready for another review.
> 
> 
> 
> I have since also tested it with a US3000C, and it works great.

Cannot wait to get it up and running

SeByDocKy commented:
Should be merge soon in the dev  version..... :) 👍

functionpointer committed:
Merge branch 'dev' into pylontech

jesserockz commented on esphome/components/pylontech/sensor/__init__.py:
```diff
@@ -0,0 +1,98 @@
+CONFIG_SCHEMA = PYLONTECH_COMPONENT_SCHEMA.extend(
+    {cv.GenerateID(): cv.declare_id(PylontechSensor)}
+    | {cv.Optional(marker): schema for marker, schema in TYPES.items()}
+)
```
```suggestion
CONFIG_SCHEMA = PYLONTECH_COMPONENT_SCHEMA.extend(
    {cv.GenerateID(): cv.declare_id(PylontechSensor)}
).extend(
    {cv.Optional(marker): schema for marker, schema in TYPES.items()}
)
```

functionpointer replied:
Unfortunately black formats it less nicely:
```python
CONFIG_SCHEMA = PYLONTECH_COMPONENT_SCHEMA.extend(
    {cv.GenerateID(): cv.declare_id(PylontechSensor)}
).extend({cv.Optional(marker): schema for marker, schema in TYPES.items()})
```
Implemented nonetheless.

jesserockz replied:
You can usually manually add commas to force black to use newlines

jesserockz commented on esphome/components/pylontech/pylontech.h:
```diff
@@ -0,0 +1,72 @@
+  }
+
+#define PYLONTECH_SENSOR(name) PYLONTECH_ENTITY_(sensor::Sensor, name)
+#define PYLONTECH_TEXT_SENSOR(name) PYLONTECH_ENTITY_(text_sensor::TextSensor, name)
```
We already have macros for these
`SUB_SENSOR(name)` in `esphome/components/sensor/sensor.h` for example
```suggestion
```

jesserockz commented on esphome/components/pylontech/pylontech.h:
```diff
@@ -0,0 +1,72 @@
+#include "esphome/core/defines.h"
+#include "esphome/components/uart/uart.h"
+
+#ifdef USE_SENSOR
+#include "esphome/components/sensor/sensor.h"
+#endif
+#ifdef USE_TEXT_SENSOR
+#include "esphome/components/text_sensor/text_sensor.h"
+#endif
```
Shouldnt be required when using listeners
```suggestion
```

jesserockz commented on esphome/components/pylontech/sensor/__init__.py:
```diff
@@ -0,0 +1,98 @@
+    bat = cg.new_Pvariable(config[CONF_ID], config[CONF_BATTERY])
+
+    for marker in TYPES:
+        if marker in config:
+            sens = await sensor.new_sensor(config[marker])
```
```suggestion
        if marker_config := config.get(marker):
            sens = await sensor.new_sensor(marker_config)
```

jesserockz commented on esphome/components/pylontech/sensor/__init__.py:
```diff
@@ -0,0 +1,98 @@
+    for marker in TYPES:
+        if marker in config:
+            sens = await sensor.new_sensor(config[marker])
+            cg.add(getattr(bat, f"set_{marker}")(sens))
```
`SUB_SENSOR` names them with a `sensor` suffix
```suggestion
            cg.add(getattr(bat, f"set_{marker}_sensor")(sens))
```

jesserockz commented on esphome/components/pylontech/__init__.py:
```diff
@@ -0,0 +1,58 @@
+
+CONF_PYLONTECH_ID = "pylontech_id"
+CONF_BATTERY = "battery"
+CONF_COULOMB = "coulomb"
+CONF_TEMPERATURE_LOW = "temperature_low"
+CONF_TEMPERATURE_HIGH = "temperature_high"
+CONF_VOLTAGE_LOW = "voltage_low"
+CONF_VOLTAGE_HIGH = "voltage_high"
+CONF_MOS_TEMPERATURE = "mos_temperature"
+
+CONF_BASE_STATE = "base_state"
+CONF_VOLTAGE_STATE = "voltage_state"
+CONF_CURRENT_STATE = "current_state"
+CONF_TEMPERATURE_STATE = "temperature_state"
+CONF_NUM_BATTERIES = "num_batteries"
```
These should move to the file that actually uses them if they are not common to multiple files, or move to the `esphome/const.py`

jesserockz commented on esphome/components/pylontech/sensor/pylontech_sensor.h:
```diff
@@ -0,0 +1,32 @@
+  void on_line_read(LineContents *line) override;
+
+ protected:
+  int bat_num_;
```
No need to assign 4 bytes for a number that is only between 1 and 6 =)
```suggestion
  uint8_t bat_num_;
```

functionpointer replied:
Good point.
I have used `int8_t` though, as it has less potential for bugs when comparing and subtracting.

jesserockz commented on esphome/components/pylontech/sensor/pylontech_sensor.h:
```diff
@@ -0,0 +1,32 @@
+
+class PylontechSensor : public PylontechListener, public Component {
+ public:
+  PylontechSensor(int bat_num);
```
```suggestion
  PylontechSensor(uint8_t bat_num);
```

jesserockz commented on esphome/components/pylontech/sensor/pylontech_sensor.h:
```diff
@@ -0,0 +1,32 @@
+  PylontechSensor(int bat_num);
+  void dump_config() override;
+
+  PYLONTECH_SENSOR(voltage)
```
```suggestion
  SUB_SENSOR(voltage)
```

jesserockz commented on esphome/components/pylontech/sensor/pylontech_sensor.cpp:
```diff
@@ -0,0 +1,60 @@
+void PylontechSensor::dump_config() {
+  ESP_LOGCONFIG(TAG, "Pylontech Sensor:");
+  ESP_LOGCONFIG(TAG, " Battery %d", this->bat_num_);
+  LOG_SENSOR("  ", "Voltage", this->voltage_);
```
```suggestion
  LOG_SENSOR("  ", "Voltage", this->voltage_sensor_);
```

jesserockz commented on esphome/components/pylontech/sensor/pylontech_sensor.cpp:
```diff
@@ -0,0 +1,60 @@
+
+static const char *const TAG = "pylontech.sensor";
+
+PylontechSensor::PylontechSensor(int bat_num) { this->bat_num_ = bat_num; }
```
```suggestion
PylontechSensor::PylontechSensor(uint8_t bat_num) { this->bat_num_ = bat_num; }
```

jesserockz commented on esphome/components/pylontech/text_sensor/__init__.py:
Please make the same changes here as in `sensor/__init.py`

functionpointer committed:
pylontech: feedback

functionpointer committed:
pylontech: remove a single newline

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

functionpointer committed:
Merge branch 'dev' into pylontech

jesserockz reviewed (dismissed):
Thanks for making those changes =)

jesserockz commented on esphome/components/pylontech/__init__.py:
```diff
@@ -0,0 +1,46 @@
+import esphome.config_validation as cv
+from esphome.const import CONF_ID
+
+CODEOWNERS = ["@functionpointer"]
+DEPENDENCIES = ["uart"]
+AUTO_LOAD = ["sensor"]
+_LOGGER = logging.getLogger(__name__)
+MULTI_CONF = True
```
Just a small tidy up and removal of auto load that is not needed
```suggestion
_LOGGER = logging.getLogger(__name__)

CODEOWNERS = ["@functionpointer"]
DEPENDENCIES = ["uart"]
MULTI_CONF = True
```

jesserockz committed:
Update esphome/components/pylontech/__init__.py

jesserockz commented on esphome/components/pylontech/__init__.py:
```diff
@@ -0,0 +1,46 @@
+
+PYLONTECH_COMPONENT_SCHEMA = cv.Schema(
+    {
+        cv.Required(CONF_PYLONTECH_ID): cv.use_id(PylontechComponent),
```
This can be used for automatic id resolving
```suggestion
        cv.GenerateID(CONF_PYLONTECH_ID): cv.use_id(PylontechComponent),
```

jesserockz committed:
Update esphome/components/pylontech/__init__.py
